File: bargeboard/providers.py
"""Per-car OpenTelemetry provider construction.

One car = one OTel Resource = one TracerProvider + MeterProvider + LoggerProvider.
Cars on the same team share `service.name` but differ on `service.instance.id`, so
axolot(e)l groups them together while still distinguishing the two drivers.

We do not register any provider globally (no `trace.set_tracer_provider`). Providers
are kept in a `ProviderBundle` per car and passed explicitly to the emitter.
"""
from __future__ import annotations

from dataclasses import dataclass
from typing import Dict

from opentelemetry.exporter.otlp.proto.grpc._log_exporter import OTLPLogExporter
from opentelemetry.exporter.otlp.proto.grpc.metric_exporter import OTLPMetricExporter
from opentelemetry.exporter.otlp.proto.grpc.trace_exporter import OTLPSpanExporter
from opentelemetry.sdk._logs import LoggerProvider
from opentelemetry.sdk._logs.export import BatchLogRecordProcessor
from opentelemetry.sdk.metrics import MeterProvider
from opentelemetry.sdk.metrics.export import PeriodicExportingMetricReader
from opentelemetry.sdk.resources import Resource
from opentelemetry.sdk.trace import TracerProvider
from opentelemetry.sdk.trace.export import BatchSpanProcessor

from bargeboard.models import DriverInfo, SessionInfo


@dataclass
class ProviderBundle:
    driver: DriverInfo
    tracer_provider: TracerProvider
    meter_provider: MeterProvider
    logger_provider: LoggerProvider

    def shutdown(self) -> None:
        """Flush and close all three providers."""
        self.tracer_provider.shutdown()
        self.meter_provider.shutdown()
        self.logger_provider.shutdown()


def make_resource(driver: DriverInfo, session: SessionInfo) -> Resource:
    return Resource.create(
        {
            "service.name": driver.team,
            "service.instance.id": driver.code,
            "f1.driver.code": driver.code,
            "f1.driver.full_name": driver.full_name,
            "f1.car.number": driver.car_number,
            "f1.team": driver.team,
            "f1.session.year": session.year,
            "f1.session.round": session.round_name,
            "f1.session.type": session.session_type,
        }
    )


def make_provider_bundle(
    driver: DriverInfo,
    session: SessionInfo,
    endpoint: str,
) -> ProviderBundle:
    resource = make_resource(driver, session)

    # Traces
    tp = TracerProvider(resource=resource)
    tp.add_span_processor(
        BatchSpanProcessor(OTLPSpanExporter(endpoint=endpoint, insecure=True))
    )

    # Metrics
    metric_reader = PeriodicExportingMetricReader(
        OTLPMetricExporter(endpoint=endpoint, insecure=True),
        export_interval_millis=1000,
    )
    mp = MeterProvider(resource=resource, metric_readers=[metric_reader])

    # Logs
    lp = LoggerProvider(resource=resource)
    lp.add_log_record_processor(
        BatchLogRecordProcessor(OTLPLogExporter(endpoint=endpoint, insecure=True))
    )

    return ProviderBundle(
        driver=driver,
        tracer_provider=tp,
        meter_provider=mp,
        logger_provider=lp,
    )


def make_provider_bundles(
    drivers: list[DriverInfo],
    session: SessionInfo,
    endpoint: str,
) -> Dict[str, ProviderBundle]:
    """Build one ProviderBundle per driver, keyed by driver code."""
    return {d.code: make_provider_bundle(d, session, endpoint) for d in drivers}

File: src/providers.ts
/**
 * Build TracerProvider + MeterProvider + LoggerProvider bundles.
 *
 * Per-driver bundles (traces + logs only) share a team `service.name` but
 * differ on `service.instance.id`. The session bundle owns the race-root
 * span and ALL metrics — every instrument lives on the single "race"
 * resource with `f1.driver.code` / `f1.team` as datapoint attributes, so
 * cross-driver queries never span resources.
 *
 * Metrics deliberately bypass the SDK MeterProvider: datapoints need
 * historical race timestamps, which the SDK can't produce. The session
 * bundle exposes the raw OTLPMetricExporter; emit/metrics.ts (MetricBank)
 * builds the datapoints.
 *
 * No providers are registered globally — every emitter takes its bundle by
 * reference. Shutdown flushes traces/logs/metrics in that order.
 */

import { OTLPLogExporter } from "@opentelemetry/exporter-logs-otlp-grpc";
import { OTLPMetricExporter } from "@opentelemetry/exporter-metrics-otlp-grpc";
import { OTLPTraceExporter } from "@opentelemetry/exporter-trace-otlp-grpc";
import {
  LoggerProvider,
  BatchLogRecordProcessor,
  ConsoleLogRecordExporter,
  SimpleLogRecordProcessor,
  type LogRecordExporter,
} from "@opentelemetry/sdk-logs";
import {
  NodeTracerProvider,
} from "@opentelemetry/sdk-trace-node";
import type { Resource } from "@opentelemetry/resources";
import {
  BatchSpanProcessor,
  ConsoleSpanExporter,
  SimpleSpanProcessor,
  type SpanExporter,
} from "@opentelemetry/sdk-trace-base";

import type { DriverInfo, SessionInfo } from "./models.js";
import { makeDriverResource, makeSessionResource } from "./resource.js";

export interface ExportTargets {
  endpoint: string;       // "host:port" for OTLP gRPC
  dryRun: boolean;        // true => no exporters, no network
  consoleEcho: boolean;   // true => add console exporters too (for inspection)
}

export interface DriverBundle {
  driver: DriverInfo;
  tracerProvider: NodeTracerProvider;
  loggerProvider: LoggerProvider;
  shutdown: () => Promise<void>;
}

export interface SessionBundle {
  session: SessionInfo;
  resource: Resource;
  tracerProvider: NodeTracerProvider;
  /** Raw OTLP metric exporter for the MetricBank (null in dry-run). The SDK
   *  MeterProvider pipeline is deliberately absent — it stamps datapoints at
   *  collection time, and we need historical race timestamps. */
  metricExporter: OTLPMetricExporter | null;
  shutdown: () => Promise<void>;
}

function makeTraceExporter(t: ExportTargets): SpanExporter | null {
  if (t.dryRun) return null;
  return new OTLPTraceExporter({ url: `http://${t.endpoint}` });
}

function makeMetricExporter(t: ExportTargets): OTLPMetricExporter | null {
  if (t.dryRun) return null;
  return new OTLPMetricExporter({ url: `http://${t.endpoint}` });
}

function makeLogExporter(t: ExportTargets): LogRecordExporter | null {
  if (t.dryRun) return null;
  return new OTLPLogExporter({ url: `http://${t.endpoint}` });
}

export function makeSessionBundle(session: SessionInfo, t: ExportTargets): SessionBundle {
  const resource = makeSessionResource(session);

  const tp = new NodeTracerProvider({ resource });
  const trExp = makeTraceExporter(t);
  if (trExp) tp.addSpanProcessor(new BatchSpanProcessor(trExp));
  if (t.consoleEcho) tp.addSpanProcessor(new SimpleSpanProcessor(new ConsoleSpanExporter()));

  const metricExporter = makeMetricExporter(t);

  return {
    session,
    resource,
    tracerProvider: tp,
    metricExporter,
    shutdown: async () => {
      await tp.shutdown();
      if (metricExporter) await metricExporter.shutdown();
    },
  };
}

export function makeDriverBundle(driver: DriverInfo, t: ExportTargets): DriverBundle {
  const resource = makeDriverResource(driver);

  const tp = new NodeTracerProvider({ resource });
  const trExp = makeTraceExporter(t);
  if (trExp) tp.addSpanProcessor(new BatchSpanProcessor(trExp));
  if (t.consoleEcho) tp.addSpanProcessor(new SimpleSpanProcessor(new ConsoleSpanExporter()));

  const lp = new LoggerProvider({ resource });
  const lExp = makeLogExporter(t);
  if (lExp) lp.addLogRecordProcessor(new BatchLogRecordProcessor(lExp));
  if (t.consoleEcho) lp.addLogRecordProcessor(new SimpleLogRecordProcessor(new ConsoleLogRecordExporter()));

  return {
    driver,
    tracerProvider: tp,
    loggerProvider: lp,
    shutdown: async () => {
      await tp.shutdown();
      await lp.shutdown();
    },
  };
}

export function makeDriverBundles(
  drivers: DriverInfo[],
  t: ExportTargets,
): Map<string, DriverBundle> {
  const out = new Map<string, DriverBundle>();
  for (const d of drivers) out.set(d.code, makeDriverBundle(d, t));
  return out;
}
